Speed up JSON.stringify by adding a separate fast case algorithm

darinadler opened the pull request:
#### e608d7dc9893576537a6f09a31e89dbecfec28e9
<pre>
Speed up JSON.stringify by adding a separate fast case algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=242286">https://bugs.webkit.org/show_bug.cgi?id=242286</a>

Reviewed by Mark Lam and Yusuke Suzuki.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
Removed ChangeLog since that file doesn&apos;t exist any more.

* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::Stringifier::stringify): Change this to be a static member function and the
main entry point for the full stringifier. Return a String instead of a JSValue.
(JSC::FastStringifier::logOutcome): Added. When FAST_STRINGIFY_LOG_USAGE is on,
collect outcomes so we can write out the histogram of reasons we were unable to
use the fast stringify.
(JSC::FastStringifier::FastStringifier): Added.
(JSC::FastStringifier::haveFailure const): Added.
(JSC::FastStringifier::result const): Added.
(JSC::FastStringifier::recordFailure): Added.
(JSC::FastStringifier::recordBufferFull): Added.
(JSC::FastStringifier::remainingCapacity const): Added.
(JSC::FastStringifier::firstGetterSetterPropertyName): Added.
(JSC::FastStringifier::recordFastPropertyEnumerationFailure): Added.
(JSC::FastStringifier::mayHaveToJSON const): Check if there may be a toJSON
function. If the prototype does not have one, we can use the fast stringifier.
(JSC::FastStringifier::append): Added.
(JSC::FastStringifier::stringify): Added.
(JSC::stringify): Try FastStringifier first, then use the full Stringifier only
if that fails.
(JSC::jsonProtoFuncStringify): Call stringify.
(JSC::JSONStringify): Call stringify.

* Source/JavaScriptCore/runtime/JSObject.h: Added
noSideEffectMayHaveNonIndexProperty.

* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::noSideEffectMayHaveNonIndexProperty): Added.

* Source/WTF/WTF.xcodeproj/project.pbxproj: Added EscapedFormsForJSON.h.
* Source/WTF/wtf/CMakeLists.txt: Ditto.

* Source/WTF/wtf/dtoa/double-conversion.h: Add WTF_EXPORT_PRIVATE so we can
use DoubleToStringConverter::ToShortest from JavaScriptCore.

* Source/WTF/wtf/text/EscapedFormsForJSON.h: Added.
* Source/WTF/wtf/text/StringBuilderJSON.cpp: Moved the escapedFormsForJSON
array into the header so it can bs used by JSONObject.cpp too.

Canonical link: <a href="https://commits.webkit.org/252326@main">https://commits.webkit.org/252326@main</a>
</pre>

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -876,11 +1369,21 @@ JSC_DEFINE_HOST_FUNCTION(jsonProtoFuncParse, (JSGlobalObject* globalObject, Call
+    }
 
-    Stringifier stringifier(globalObject, callFrame->argument(1), callFrame->argument(2));
+    auto scope = DECLARE_THROW_SCOPE(vm);
```
We want to keep this DECLARE_THROW_SCOPE at the top of the function like before.  For Release builds, it will be effectively a no-op.  For Debug builds, we want it at the top because that way, it will simulate a potential throw for exception check validation even if we never take the slow path in our test workloads.

darinadler replied:
OK, I will move it to the top. Will I have to do anything with it inside the `if` statement?

darinadler replied:
Took care of this.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,490 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+{
+    VM& vm = globalObject.vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    PropertyNameArray names(globalObject.vm(), PropertyNameMode::Strings, PrivateSymbolMode::Include);
```
nit: just use vm instead of globalObject.vm()

darinadler replied:
Thanks. Fixed locally.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,490 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    if (!function) {
+        auto scope = DECLARE_THROW_SCOPE(m_vm);
+        scope.assertNoException();
+        PropertySlot slot(&object, PropertySlot::InternalMethodType::Get);
```
I don't think using PropertySlot::InternalMethodType::Get is side effect free.  We normally use PropertySlot::InternalMethodType::VMInquiry when we don't want side effects.  However, I think VMInquiry will result in getPropertySlot returning false if there's a potential for side effects, which is the opposite of what you want.

@Constellation can you double check my understanding here?

Constellation replied:
Yes, `PropertySlot::InternalMethodType::Get` is normal `[[Get]]` operation in JS, thus, it can have a side effect (calling getters etc.).

Constellation replied:
Ah, but we are just using `object.getPropertySlot`. So we do not have a side effect.

darinadler replied:
I am pretty sure that the current code **does** have side effects, when calling `slot.getValue`. For the purposes here of answering the question “might have `tojSON`”, we don’t actually have to get the value; the value is only needed for the caching. I think I can fix the problem by just returning true and not caching anything when hasProperty is true and removing the call to `slot.getValue`.

darinadler replied:
I got rid of the call to getValue.

MenloDorian reviewed (commented):
I'm still reading the patch, but thought I'd post my comments so far.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,492 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    scope.assertNoException();
+    PropertySlot slot(&object, PropertySlot::InternalMethodType::Get);
+    bool hasProperty = object.getPropertySlot(&m_globalObject, m_vm.propertyNames->toJSON, slot);
+    CLEAR_AND_RETURN_IF_EXCEPTION(scope, true);
```
My understanding is that FastStringifier must not execute side effects.  Otherwise, the side effect will have executed twice when we execute the slow path.

However, AFAICT, object.getPropertySlot itself can execute side effects e.g.
    JSObject::getPropertySlot(JSGlobalObject* globalObject, PropertyName propertyName, PropertySlot& slot)
        calls JSObject::getNonIndexPropertySlot if TypeInfo::overridesGetOwnPropertySlot()
    JSObject::getNonIndexPropertySlot(JSGlobalObject* globalObject, PropertyName propertyName, PropertySlot& slot)
        calls structure->classInfoForCells()->methodTable.getOwnPropertySlot(object, globalObject, propertyName, slot)
        which can call anything with side effects e.g.
            ProxyObject::getOwnPropertySlot, which calls
            ProxyObject::getOwnPropertySlotCommon, which calls
            ProxyObject::performGet, which calls
            ProxyObject::performGet, which calls an arbitrary JS callback.

I think for what you need here, you'll have to write a new JSObject method based on JSObject::getPropertySlot e.g. 
    bool JSObject::noSideEffectMayHaveProperty(JSGlobalObject* globalObject, PropertyName propertyName);

... which conservatively returns true if it might execute side effects or definitely find the specified property.  For example, if TypeInfo::overridesGetOwnPropertySlot() is true, conservatively return true.  I don't know if this is too conservative, and therefore pessimistic.

With that, you will not need the CLEAR_AND_RETURN_IF_EXCEPTION() because it should never need to throw an exception (or do any exception checks in its implementation).  Hmmm, looks like you might not even need the ThrowScope.  Note: any place that does an exception check can inadvertently throw a TerminationException (which can be requested asynchronously) unless it is guarded within a DeferTerminationForAWhile or DeferTermination scope.

Constellation replied:
Ah, I misunderstood. If this function needs to be side-effect free, then we cannot use `getPropertySlot` with `Get` here.
Originally these code exists for normal `toJSON` property path, and it should have side-effect. So it was no problem.

darinadler replied:
If we must have a new `noSideEffectMayHaveProperty` function, then that’s probably the biggest single thing I’d need to do for this patch. Everything else is entirely confined to `JSONObject.cpp`. Maybe someone can help me figure out how to do that correctly, but it’s harder for me than everything else I did so far.

And yes, the possibility that `overridesGetOwnPropertySlot` will exclude too many objects is something I’ll have to test for; I suspect it won’t be a problem. The reason that `FAST_STRINGIFY_LOG_USAGE` exists is to check on things like this.

I guess it’s obvious to me in hindsight that any path where there may be an exception is also one where we may have side effects, so any use of `CLEAR_AND_RETURN_IF_EXCEPTION` is a sign that the code is wrong.

darinadler replied:
I fixed this, adding `noSideEffectMayHaveProperty`.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,492 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        // Getting the property value so we can cache it could cause side effects; instead return true without caching.
+        RELEASE_AND_RETURN(scope, true);
+    }
+    CLEAR_AND_RETURN_IF_EXCEPTION(scope, true);
```
This redundant, and is not needed.  All the more so if you implement and use a JSObject::noSideEffectMayHaveProperty.

darinadler replied:
Understood; will fix.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -286,6 +292,9 @@ JSValue Stringifier::stringify(JSValue value)
     if (UNLIKELY(stringifyResult != StringifySucceeded))
         return jsUndefined();
+#if FAST_STRINGIFY_LOG_USAGE
+    WTFLogAlways("Not fastStringify: %s", result.toString().utf8().data());
```
JSC uses `dataLogLn("Not fastStringify: ", result.toString());`

darinadler replied:
I am 100% willing to use `dataLogLn`; I will fix this right away.

Separately, I am deeply offended by the concept "JSC uses". Why should this be different in JavaScriptCore than in any other framework built on top of WTF? Is there some special requirement JavaScriptCore has that WebCore and WebKit do not?

MenloDorian replied:
`dataLog` supports custom dumpers, which is why we use it.  dataLog gets printed to stderr by default.  For JSC folks, we work with the jsc shell, and we see the `dataLog` output.  For WebCore/WebKit, stderr is not always dumped to the terminal.  Hence, they rely on `WTFLogAlways` so that they can see the logging via syslog.  However, we now have a `JSC_useOSLog=1` option that will redirect everything printed via dataLog to syslog.  So, technically, WebCore and WebKit can use it too.

darinadler replied:
OK, but still not happy with "technically". We should use one thing across the project. I am annoyed that we aren’t even trying to do that!

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,492 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        std::sort(vector.begin(), vector.end(), [](auto& a, auto &b) {
+            return a.value != b.value ? a.value > b.value : codePointCompareLessThan(a.key, b.key);
+        });
+        WTFLogAlways("fastStringify outcomes");
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,492 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        });
+        WTFLogAlways("fastStringify outcomes");
+        for (auto& pair : vector)
+            WTFLogAlways("%4u: %s", pair.value, pair.key.utf8().data());
```
Ditto.

MenloDorian replied:
If you want the nice printf style formatting, you can use dataLogF.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +636,492 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    if (size <= maxSizeSeen)
+        return;
+    maxSizeSeen = size;
+    WTFLogAlways("max fastStringify buffer size used: %u", size);
```
Ditto.

darinadler commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    return std::size(m_buffer) - m_length;
+}
+
+#if FAST_STRINGIFY_LOG_USAGE && !FAST_STRINGIFY_GETTER_EXPERIMENT
```
Oops, should remove the `&& !FAST_STRINGIFY_GETTER_EXPERIMENT` here, that’s a mistake. But it only has an effect when people turn on `FAST_STRINGIFY_GETTER_EXPERIMENT`. Maybe I should just remove that flag and experiment entirely before landing; not sure.

darinadler replied:
Fixed locally.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    }
+
+    if (!value.isCell()) {
+        recordFailure("value type"_s);
```
Does `isDouble()` worth handling? Maybe a bit too complicated to compute the length.

darinadler replied:
We don’t hit the `isDouble` case in Speedometer. I would be prefer not to add it now. And would be happy to add it later. Not super-complicated to get the length; would just do the operation into a `NumberToStringBuffer` and then `memcpy` the results if it fits.

darinadler replied:
To be fair, I don’t think we hit the null, true, or false cases either, so my thinking on this is not 100% logical.

MenloDorian replied:
Are we sure if the `isDouble` case is actually rare with in the wild usage of JSON that get stringify?  If not, then this would be imposing a perf penalty on those cases just because they don't match Speedometer's profile.  Our goal shouldn't be just to make Speedometer fast, but to have a fast `JSON.stringify`.  So, unless there's some unwanted perf cost for adding the `isDouble` case, I feel that we should add it especially since it's not hard to do.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                }
+                bool needComma = m_buffer[m_length - 1] != '{';
+                unsigned nameLength = name.length();
+                if (remainingCapacity() < needComma + 1 + nameLength + 2) {
```
BTW, String.length()'s maximum value is INT32_MAX, so, this never overflows. (so this is fine).

darinadler replied:
Yes, I did think about that when writing the code. Decided not to add an explicit comment about it. Thank you for reaffirming that it’s OK.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        if (object.structureID() != structure.id()) {
+            recordFailure("structure ID mismatch"_s);
+            return;
+        }
```
We do not need to have this check. This Structure check was existing only to capture structure transition from the previous operations.

darinadler replied:
Thanks, removed locally.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                scope.clearException();
+                recordFailure("getOwnPropertyNames exception"_s);
+                return;
+            }
```
I think we need to keep this error. It is throwing an error and it is observable because `getOwnPropertyNames` can have side-effect. So, we need to throw this error (exact this error) to users.
So, let's not clear exception, and capture it in the root caller of `FastStringifier`, and just throwing an error.

darinadler replied:
Since this is part of my getter-side-effects experiment I think I will just remove the code. Later we can see if we can come up with a side-effect-free way to run trivial getters, but for now I will set it aside.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        if (array.isCallable()) {
+            recordFailure("callable array"_s);
+            return;
+        }
```
Let's remove this. Array is never callable. We can add `ASSERT(!array.isCallable())`

darinadler replied:
Removed it. I am not going to add the assert, though.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                return false;
+            }
+            return true;
+        });
```
Here is a tip. Most of arrays do not have their own custom property. So, let's add a fast path,

```
if (m_globalObject.isOriginalArrayStructure(array.structure())) {
    // Then, there is no toJSON property. This is original array structure, which means, no custom properties are added.
} else {
    // Check toJSON in the own properties.
}
```

darinadler replied:
Great idea. Will do.

darinadler replied:
Done.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                m_buffer[m_length + 1 + nameLength] = '"';
+                m_buffer[m_length + 1 + nameLength + 1] = ':';
+                m_length += 1 + nameLength + 2;
+                append(object.getDirect(entry.offset()));
```
Even though we do not see non-user-observable behavior as no-side-effect, still it is possible that structure can be transitioned. For example, `append` will call `getOwnPropertyNames`, and it could change the state of this object.
To avoid crash (like, object's structure could be flattened, and in that case, it is possible that offset->value mapping is completely changed in the middle of iteration), we should buffer values in some places, and use them.
See `Object.assign`'s implementation for more detail.

darinadler replied:
Append will not call `getOwnPropertyNames`. That was a disabled experiment and I have now deleted it in the new version of the patch. Could you clarify? If I remove the `FAST_STRINGIFY_GETTER_EXPERIMENT` code is this buffering still needed?

darinadler replied:
I’m assuming the answer is no. Please let me know if I am wrong!

darinadler replied:
We’re discussing this below, so I will resolve this one.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                m_buffer[m_length + 1 + nameLength] = '"';
+                m_buffer[m_length + 1 + nameLength + 1] = ':';
+                m_length += 1 + nameLength + 2;
+                scope.assertNoException();
```
Let's remove this. Unfortunately, this is some kind of stale function since exception can happen at any time if there is scope (termination exception can be thrown from the watchdog thread asynchronously.)

MenloDorian replied:
If you don't want interference from the `TerminationException`, declare a `DeferTerminationForAWhile` scope at top of your scope e.g. in `fastStringify`.  This is ok to do if your function is bounded i.e. it won't be calling arbitrary code (like calling back into JS) and is guaranteed to terminate in some smallish finite time in the near future, thereby letting someone else service the termination request later instead of you.

That said, you're in `FastStringifier`, and I would expect side-effect free code to never throw, and therefore, won't need exception checks.  No exception checks == no chance for a termination request to be serviced, hence, should not see a `TerminationException`.

MenloDorian replied:
Obviously, the patch is doing exception checks (though I haven't grok why this is needed yet).  So TerminationExceptions are possible.

darinadler replied:
This is all in the experimental disabled code. I will resolve this by deleting that code from my patch.

Constellation commented on Source/WTF/wtf/text/EscapedFormsForJSON.h:
```diff
@@ -0,0 +1,67 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
```
Let's copy SpiderMonkey's license header in `StringBuilderJSON.cpp` (That's the main reason why we have separate `StringBuilderJSON.cpp` from `StringBuilder.cpp`.

darinadler replied:
Done.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                    scope.clearException();
+                    recordFailure("get exception"_s);
+                    return;
+                }
```
Do we have a side-effect on this? For example, if there is a getter, calling this can be user-observable. So, if

1. First use FastJSONStringfier
2. Fail
3. Use generic Stringifier

Then, user can observe (1) and (2) since getter is called twice.

darinadler replied:
Yes, this was part of the deleted experiment. That’s why it was disabled.

Constellation reviewed (commented):
Fun!

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+#else
+            auto scope = DECLARE_THROW_SCOPE(m_vm);
+            PropertyNameArray names(m_vm, PropertyNameMode::Strings, PrivateSymbolMode::Exclude);
+            object.methodTable()->getOwnPropertyNames(&object, &m_globalObject, names, DontEnumPropertiesMode::Exclude);
```
By checking `OverridesGetOwnPropertyNames` bit in object's typeinfo, you can know that there is no special customized getOwnPropertyNames for this object. So probably, we should check it and if there is the one, then let's bail out.

darinadler replied:
Part of the disabled, now-deleted experiment.

MenloDorian commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -206,6 +206,32 @@ inline JSValue JSObject::getIfPropertyExists(JSGlobalObject* globalObject, Prope
+    ASSERT(propertyName != vm.propertyNames->length);
+    for (auto* object = this; object; object = object->getPrototypeDirect().getObject()) {
+        auto inlineTypeFlags = object->inlineTypeFlags();
+        if (TypeInfo::overridesGetOwnPropertySlot(inlineTypeFlags) && object->classInfo() != ArrayPrototype::info())
```
Would you like to exclude JSArray::info() as well?

darinadler replied:
I am willing to add that in the future if it’s helpful. For the moment, there is no need for that additional check.

This is part of the tension for this function; is it general purpose or just for one particular use in `fastStringify`? At the moment there is no real benefit to adding a check for `JSArray` given how this is used.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+{
+    if (haveFailure())
+        return { };
+    addStringifyBufferSize(m_length);
```
nit: calling this function `addStringBufferSize` seems strange because it isn't really adding anything together (which is what it suggested to me when I first read it).  What do you think of calling it `recordStringBufferSize`?  Ditto for `addStringifyOutcome` => `recordStringifyOutcome`.  This also makes them consistent with `recordFailure` and `recordBufferFull` below: all are recording a result / outcome.

darinadler replied:
Yes, good suggestion. I will rename them.

darinadler replied:
Fixed locally.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        auto* characters = string.characters8();
+        for (unsigned i = 0; i < stringLength; ++i) {
+            auto character = characters[i];
+            if (WTF::escapedFormsForJSON[character]) {
```
Should we make all these fail out checks in `append` use `UNLIKELY` so that the success case will have greater code locality?  The premise of this fast path is that we expect it to succeed most of the time.  And when it fails, we only need to execute one of the fail paths.  So, I think it makes sense to give the success path a chance to be more co-located.

MenloDorian replied:
To clarify, I meant that the expectation that the fast path will succeed most of the time implies that the fail out cases are unlikely.  Hence, it may be profitable to apply `UNLIKELY` to the fail out conditions.

darinadler replied:
I added lots of `UNLIKELY`. Still not sure it was the right thing to do.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                recordFailure("wrapped big int"_s);
+                return;
+            }
+        } while ((ancestorClass = ancestorClass->parentClass));
```
`NumberObject`, `StringObject`, and `BooleanObject`'s `JSType` are not `ObjectType` nor `FinalObjectType`.  Only `BigIntObject`'s `JSType` is `ObjectType`.  Hence, you can change this into a while loop instead of a do-while, and save on the first iteration.  However, you'll have to add an explicit check for the `BigIntObject` case before you enter the while loop.

darinadler replied:
That’s great, thank you! I will remove the `NumberObject`, `StringObject` and `BooleanObject` checks. Not sure I want to change this into a while loop, though. It would be elegant as a for loop, but that would add one unnecessary null check. For now I think I will leave it as a do/while.

MenloDorian replied:
Oh, that's right: the parentClass relationship is static in our code base.  A quick grep reveals that only `NumberPrototype`, `StringPrototype`, and `BooleanPrototype` has these 3 classInfo as their parentClass.  But all 3 of those also do not have `JSType`s of `ObjectType` nor `FinalObjectType`.  Hence, it means that we don't need to check for these when iterating the parentClass chain.

That said, there is also no classInfo that has `BigIntObject::s_info` as its parentClass.  So, you don't really need a loop here.  You only need to check if the top level classInfo is `BigIntObject::s_info`.

darinadler replied:
I wonder if `BigIntObject` will pass the `canPerformFastPropertyEnumerationForJSONStringify` check. If it won’t, then I don’t need the `BigIntObject` check at all!

darinadler replied:
There are many other checks, and if any are guaranteed false for `BigIntObject` then we can remove the check. For example, if the prototype is guaranteed to not be the default object prototype, or if `canPerformFastPropertyEnumerationForJSONStringify` will return false.

darinadler replied:
OK, big int wrapped values have the `BigIntPrototype`, which is not the standard object prototype, so we don't need the `BigIntObject` check at all.

MenloDorian replied:
BigIntObject does use the default object prototype.  AFAICT, it also doesn't set its own StructureFlags.  Hence, I think `canPerformFastPropertyEnumerationForJSONStringify` will return true.

MenloDorian replied:
Oh you are correct.  I was checking the `CREATE_PROTOTYPE_FOR_SIMPLE_TYPE` macro, and conflated `m_bigIntPrototype`'s base prototype (which is the default object prototype) for `m_bigIntPrototype` itself.  So yes, the `BigIntObject` check is not needed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+            recordFailure("callable object"_s);
+            return;
+        }
+        auto* ancestorClass = object.classInfo();
```
Fetching `classInfo()` entails first fetching `structure()` in its implementation.  Consider pre-fetching `structure()` explicitly here (so that you can also use it after the loop below), and then fetch the classInfo  using `structure->classInfoForCells()`.  The computation of structure involves multiple instructions (though not slow), and it looks like you should have enough registers for the compiler to keep structure in a register.

darinadler replied:
Done. Slightly sad that doing this causes me to lose the assertions from `JSCell::classInfo()`, but not so sad that I will add them.

MenloDorian reviewed (approved):
r=me with suggestions.

Constellation commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -627,6 +635,456 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+            m_buffer[m_length + 1 + nameLength] = '"';
+            m_buffer[m_length + 1 + nameLength + 1] = ':';
+            m_length += 1 + nameLength + 2;
+            append(object.getDirect(entry.offset()));
```
I think still it is possible that non-user-observable behavior can change the structure. Like, reify of properties, custom values etc.
So, let's add a check,
```
if (UNLIKELY(&structure != object.structure()))
    // FAIL
```
Even though any kind of side-effect for structure happens, if structure is not changed, then it is OK.

darinadler replied:
I am **absolutely** willing to add this check if it’s needed; I will need to re-run performance testing, but I bet it won’t hurt too much. But I still don’t understand. Is it `getDirect` that can reify properties or `forEachProperty`? Can I avoid this by using `propertyTableOrNull` instead?

Constellation replied:
No, what I'm worrying is some operations in `append(...)`. Do we have any guarantee that some code in `append` will not invoke the structure transition?

darinadler replied:
This function itself is `append`; it’s a recursive call to this function. We can look over the whole function to see what is risky and could cause a structure transition. To me it seems obvious that nothing `append` does can cause a transition, but maybe that’s because I don’t know what to look for.

darinadler replied:
By the way, it would be OK if it causes a transition and a failure, because if we have a failure it immediately exits. If it can cause a transition and succeed then, yes, that would it be a problem.

As I said, I have the humility to admit there could be a transition because my understanding isn’t deep enough to appreciate something we are doing here in `append` that could have that subtle effect, but I **don’t** think there’s anything here that can do that.

Constellation replied:
If we missed it, it becomes serious security issue since it allows attackers to access to non-used place with newly transitioned (sometimes, it gets flattened) structure. Historically, there are many missed cases (e.g. Object.assign's reify case). And since structure transition is not user-observable behavior, there is no guarantee that future code will perform structure transition, which is used in somewhere in `append` function, and we cannot notice that this place becomes a security issue after that.
So, so long as we have 100% confident & we have a mechanism to assert this guarantee, I think we should check it.
Compared to user-observable-behaviors, JSC developers do not see structure transitions as a side-effect in the code, so I cannot guarantee that `append` can / will not include that code.

darinadler replied:
OK, I can put in an assertion and a check from a general sense of security vulnerability paranoia/hardening

darinadler replied:
Is the pointer check sufficient? Can the old structure’s memory be deallocated as part of a structure transition and possible be reused for a new structure?

darinadler replied:
New patch uploaded has this check added.

darinadler committed:
Speed up JSON.stringify by adding a separate fast case algorithm
https://bugs.webkit.org/show_bug.cgi?id=242286

Reviewed by Mark Lam and Yusuke Suzuki.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
Removed ChangeLog since that file doesn't exist any more.

* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::Stringifier::stringify): Change this to be a static member function and the
main entry point for the full stringifier. Return a String instead of a JSValue.
(JSC::FastStringifier::logOutcome): Added. When FAST_STRINGIFY_LOG_USAGE is on,
collect outcomes so we can write out the histogram of reasons we were unable to
use the fast stringify.
(JSC::FastStringifier::FastStringifier): Added.
(JSC::FastStringifier::haveFailure const): Added.
(JSC::FastStringifier::result const): Added.
(JSC::FastStringifier::recordFailure): Added.
(JSC::FastStringifier::recordBufferFull): Added.
(JSC::FastStringifier::remainingCapacity const): Added.
(JSC::FastStringifier::firstGetterSetterPropertyName): Added.
(JSC::FastStringifier::recordFastPropertyEnumerationFailure): Added.
(JSC::FastStringifier::mayHaveToJSON const): Check if there may be a toJSON
function. If the prototype does not have one, we can use the fast stringifier.
(JSC::FastStringifier::append): Added.
(JSC::FastStringifier::stringify): Added.
(JSC::stringify): Try FastStringifier first, then use the full Stringifier only
if that fails.
(JSC::jsonProtoFuncStringify): Call stringify.
(JSC::JSONStringify): Call stringify.

* Source/JavaScriptCore/runtime/JSObject.h: Added
noSideEffectMayHaveNonIndexProperty.

* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::noSideEffectMayHaveNonIndexProperty): Added.

* Source/WTF/WTF.xcodeproj/project.pbxproj: Added EscapedFormsForJSON.h.
* Source/WTF/wtf/CMakeLists.txt: Ditto.

* Source/WTF/wtf/dtoa/double-conversion.h: Add WTF_EXPORT_PRIVATE so we can
use DoubleToStringConverter::ToShortest from JavaScriptCore.

* Source/WTF/wtf/text/EscapedFormsForJSON.h: Added.
* Source/WTF/wtf/text/StringBuilderJSON.cpp: Moved the escapedFormsForJSON
array into the header so it can bs used by JSONObject.cpp too.

Canonical link: https://commits.webkit.org/252326@main
